From 4ce8a36a76c186564c778667c2f68efa9949f556 Mon Sep 17 00:00:00 2001
From: Axel Huebl <axel.huebl@plasma.ninja>
Date: Fri, 5 Mar 2021 17:56:32 -0800
Subject: [PATCH 1/4] GNU extension: __attribute__((format))

Wrap this into a macro since it's a GNU extension.

Breaks MSVC.
---
 include/blas/util.hh | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/include/blas/util.hh b/include/blas/util.hh
index 40b97de..2010c6d 100644
--- a/include/blas/util.hh
+++ b/include/blas/util.hh
@@ -415,13 +415,19 @@ inline void throw_if( bool cond, const char* condstr, const char* func )
     }
 }
 
+#if defined(_MSC_VER)
+#   define BLASPP_ATTR_FORMAT(I, F)
+#else
+#   define BLASPP_ATTR_FORMAT(I, F) __attribute__((format( printf, I, F )))
+#endif
+
 // -----------------------------------------------------------------------------
 // internal helper function; throws Error if cond is true
 // uses printf-style format for error message
 // called by blas_error_if_msg macro
 // condstr is ignored, but differentiates this from other version.
 inline void throw_if( bool cond, const char* condstr, const char* func, const char* format, ... )
-    __attribute__((format( printf, 4, 5 )));
+    BLASPP_ATTR_FORMAT(4, 5);
 
 inline void throw_if( bool cond, const char* condstr, const char* func, const char* format, ... )
 {
@@ -439,7 +445,7 @@ inline void throw_if( bool cond, const char* condstr, const char* func, const ch
 // uses printf-style format for error message
 // called by blas_error_if_msg macro
 inline void abort_if( bool cond, const char* func,  const char* format, ... )
-    __attribute__((format( printf, 3, 4 )));
+    BLASPP_ATTR_FORMAT(3, 4);
 
 inline void abort_if( bool cond, const char* func,  const char* format, ... )
 {
@@ -454,6 +460,8 @@ inline void abort_if( bool cond, const char* func,  const char* format, ... )
     }
 }
 
+#undef BLASPP_ATTR_FORMAT
+
 }  // namespace internal
 
 // -----------------------------------------------------------------------------
-- 
2.25.1

